src/Bicep.Core/Semantics/ParamsSemanticModel.cs

```diff
@@ -27,11 +30,96 @@ public ParamsSemanticModel(BicepParamFile bicepParamFile, Compilation? bicepComp
+        }
+
+        private void AddMissingErrors(Metadata.ParameterMetadata[] sortedRequiredParameters, ParameterAssignmentSyntax[] sortedParameterAssignments)
+        {
+            var i = 0;
+            var j = 0;
+
+            while ((i < sortedRequiredParameters.Length) && (j < sortedParameterAssignments.Length))
+            {
+                var bicepParam = sortedRequiredParameters[i];
+                var paramParam = this.ParamBinder.GetSymbolInfo(sortedParameterAssignments[j]);
+                
+                if (LanguageConstants.IdentifierComparer.Equals(bicepParam.Name, paramParam?.Name))
+                {
+                    i++;
+                    j++;
+                }
+                else if (LanguageConstants.IdentifierComparer.Compare(bicepParam.Name, paramParam?.Name) < 0)
+                {
+                    if (bicepParam.IsRequired)
+                    {
+                        semanticDiagnostics.Add(DiagnosticBuilder.ForPosition(new TextSpan(0,0)).MissingParameterAssignment(bicepParam.Name));
+                    }
+                    i++;
+                }
+                else
+                {
+                    semanticDiagnostics.Add(DiagnosticBuilder.ForPosition(sortedParameterAssignments[j].Span).MissingParameterDeclaration(paramParam?.Name));
+                    j++;
+                }
+            }
+
+            // if i or j is at the end, emit errors for all remaining unpaired symbols 
+            while (i < sortedRequiredParameters.Count())
+            {
+                if (sortedRequiredParameters[i].IsRequired)
+                {
+                    semanticDiagnostics.Add(DiagnosticBuilder.ForPosition(new TextSpan(0,0)).MissingParameterAssignment(sortedRequiredParameters[i].Name));
+                }
+                i++;
+            }
+            while (j < sortedParameterAssignments.Count())
+            {
+                semanticDiagnostics.Add(DiagnosticBuilder.ForPosition(sortedParameterAssignments[j].Span).MissingParameterDeclaration(this.ParamBinder.GetSymbolInfo(sortedParameterAssignments[j])?.Name));
+                j++;
+            }
+        }
         
+        private void AddTypeMismatchErrors(IOrderedEnumerable<ParameterAssignmentSyntax> sortedParameterAssignmentSyntax)
+        {
+            foreach (var syntax in sortedParameterAssignmentSyntax)
+            {
+                if ((ParamsTypeManager.GetTypeInfo(syntax) is not ErrorType) && (ParamsTypeManager.GetDeclaredType(syntax) is { } declaredType) &&
+                    (!TypeValidator.AreTypesAssignable(ParamsTypeManager.GetTypeInfo(syntax), declaredType)))
+                {
+                    semanticDiagnostics.Add(DiagnosticBuilder.ForPosition(syntax.Span).TypeMismatch(this.ParamBinder.GetSymbolInfo(syntax)?.Name, declaredType, ParamsTypeManager.GetTypeInfo(syntax)));
+                }
+            }
+        }
+        
+        public IEnumerable<IDiagnostic> GetDiagnostics()
```

Comment:
I would recommend wrapping all of the diagnostics in a `Lazy<T>` - that'll ensure you only calculate diagnostics at most once per compilation. See https://github.com/Azure/bicep/blob/8e74bc9e42bc65c57cfd4518c321e3ca125e015a/src/Bicep.Core/Semantics/SemanticModel.cs#L70-L71 for an example.

Reply:
+1